cpp/src/io/orc/orc.cpp

```diff
@@ -402,6 +360,181 @@ const uint8_t *OrcDecompressor::Decompress(const uint8_t *srcBytes, size_t srcLe
   return m_buf.data();
 }
 
+metadata::metadata(datasource *const src) : source(src)
+{
+  const auto len         = source->size();
+  const auto max_ps_size = std::min(len, static_cast<size_t>(256));
+
+  // Read uncompressed postscript section (max 255 bytes + 1 byte for length)
+  auto buffer            = source->host_read(len - max_ps_size, max_ps_size);
+  const size_t ps_length = buffer->data()[max_ps_size - 1];
+  const uint8_t *ps_data = &buffer->data()[max_ps_size - ps_length - 1];
+  ProtobufReader(ps_data, ps_length).read(ps);
+  CUDF_EXPECTS(ps.footerLength + ps_length < len, "Invalid footer length");
+
+  // If compression is used, the rest of the metadata is compressed
+  // If no compressed is used, the decompressor is simply a pass-through
+  decompressor = std::make_unique<OrcDecompressor>(ps.compression, ps.compressionBlockSize);
+
+  // Read compressed filefooter section
+  buffer           = source->host_read(len - ps_length - 1 - ps.footerLength, ps.footerLength);
+  size_t ff_length = 0;
+  auto ff_data     = decompressor->Decompress(buffer->data(), ps.footerLength, &ff_length);
+  ProtobufReader(ff_data, ff_length).read(ff);
+  CUDF_EXPECTS(get_num_columns() > 0, "No columns found");
+
+  // Read compressed metadata section
+  buffer =
+    source->host_read(len - ps_length - 1 - ps.footerLength - ps.metadataLength, ps.metadataLength);
+  size_t md_length = 0;
+  auto md_data     = decompressor->Decompress(buffer->data(), ps.metadataLength, &md_length);
+  orc::ProtobufReader(md_data, md_length).read(md);
+}
+
+std::vector<metadata::OrcStripeInfo> metadata::select_stripes(const std::vector<size_type> &stripes,
+                                                              size_type &row_start,
+                                                              size_type &row_count)
+{
+  std::vector<OrcStripeInfo> selection;
+
+  if (!stripes.empty()) {
+    size_t stripe_rows = 0;
+    for (const auto &stripe_idx : stripes) {
+      CUDF_EXPECTS(stripe_idx >= 0 && stripe_idx < get_num_stripes(), "Invalid stripe index");
+      selection.emplace_back(&ff.stripes[stripe_idx], nullptr);
+      stripe_rows += ff.stripes[stripe_idx].numberOfRows;
+    }
+    // row_start is 0 if stripes are set. If this is not true anymore, then
+    // row_start needs to be subtracted to get the correct row_count
+    CUDF_EXPECTS(row_start == 0, "Start row index should be 0");
+    row_count = static_cast<size_type>(stripe_rows);
+  } else {
+    row_start = std::max(row_start, 0);
+    if (row_count < 0) {
+      row_count = static_cast<size_type>(
+        std::min<size_t>(get_total_rows() - row_start, std::numeric_limits<size_type>::max()));
+    } else {
+      row_count = static_cast<size_type>(std::min<size_t>(get_total_rows() - row_start, row_count));
+    }
```

Comment:
Could be I+LE

Reply:
Unfortunately, here we are modifying an in/out parameter so I don't think I+LE would be useful. This is again related to output parameters - if we returned the actual row count instead, we could use I+LE to set the value here.